Add ProfileEvents for number of rows and bytes passing filter transform

onkar opened the pull request:
This PR adds profile events to calculate the number of rows and number of bytes in filter transform of a query. It resolves https://github.com/ClickHouse/ClickHouse/issues/67288.


### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`FilterTransformPassedRows` and `FilterTransformPassedBytes` profile events will show the number of rows and number of bytes filtered during the query execution.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

onkar committed:
Add ProfileEvents for number of rows and bytes passing WHERE clause

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=76662) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=76662) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=76662) it.</sub>

antaljanosbenjamin commented on src/Common/ProfileEvents.cpp:
```diff
@@ -953,6 +953,9 @@ The server successfully detected this situation and will download merged part fr
     M(MemoryWorkerRunElapsedMicroseconds, "Total time spent by MemoryWorker for background work", ValueType::Microseconds) \
     \
     M(ParquetFetchWaitTimeMicroseconds, "Time of waiting fetching parquet data", ValueType::Microseconds) \
+    M(WhereFilterPassedRows, "Number of rows that passed the WHERE filter in the query", ValueType::Number) \
+    M(WhereFilterPassedBytes, "Number of bytes that passed the WHERE filter in the query", ValueType::Number) \
+
```
I am not sure if using the `WHERE` clause makes sense here. In query execution we have the `FilterTransform`, which might come from:
1. `WHERE` clause
2. `HAVING` clause
2. `JOIN ON` clause

Maybe rename them to `FilterTransformPassedRows` and `FilterTransformPassedBytes

```suggestion
    M(FilterTransformPassedRows, "Number of rows that passed the WHERE filter in the query", ValueType::Number) \
    M(FilterTransformPassedBytes, "Number of bytes that passed the WHERE filter in the query", ValueType::Bytes) \
```

onkar replied:
Done

antaljanosbenjamin commented on src/Processors/Transforms/FilterTransform.cpp:
```diff
@@ -182,6 +188,14 @@ void FilterTransform::doTransform(Chunk & chunk)
     else
         num_filtered_rows = filter_description->countBytesInFilter();
 
+    ProfileEvents::increment(ProfileEvents::WhereFilterPassedRows, num_filtered_rows);
```
We also return above when `constant_filter_description.always_true` is true, so this might not be the single place where you have to update this metric.

onkar replied:
Done

antaljanosbenjamin reviewed (commented):
We also need tests for this.

onkar committed:
Address review comments and add test

onkar committed:
Check style fix

onkar committed:
More style fixes

onkar commented:
Hi @antaljanosbenjamin, thank you for reviewing the PR. I addressed your comments, please take a look and let me know if anything else needs to be changed.

The following tests are failing:

1. [test_odbc_interaction/test.py::test_postgres_odbc_hashed_dictionary_no_tty_pipe_overflow](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=76662&sha=4af99ac1921cda1efd0508d599558acfa5d63966&name_0=PR&name_1=Integration+tests+%28release%2C+4%2F4%29&name_2=test_odbc_interaction%2Ftest.py%3A%3Atest_postgres_odbc_hashed_dictionary_no_tty_pipe_overflow) but my change has not touched any ODBC logic. 
2. [Stress test (ubsan) ](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=76662&sha=latest&name_0=PR&name_1=Stress+test+%28ubsan%29&name_2=Fatal+message+in+clickhouse-server.log+%28see+fatal_messages.txt%29) but the stack trace of this failure does not seem to be related to my change. 

Please advise on the next steps to deal with these test failures.

antaljanosbenjamin commented on src/Processors/Transforms/FilterTransform.cpp:
```diff
@@ -121,6 +127,16 @@ void FilterTransform::doTransform(Chunk & chunk)
 
     if (constant_filter_description.always_true || on_totals)
     {
+        ProfileEvents::increment(ProfileEvents::FilterTransformPassedRows, num_rows_before_filtration);
+
+        size_t num_bytes_before_filtration = 0;
+        for (const auto & column : columns)
+        {
+            if (column)
+                num_bytes_before_filtration += column->byteSize();
+        }
+        ProfileEvents::increment(ProfileEvents::FilterTransformPassedBytes, num_bytes_before_filtration);
+
```
Would you mind creating a lambda from this snipped and reuse it? I think the complexity of this few rows reached the point when it makes sense to convert it into a lambda.

onkar replied:
Done.

antaljanosbenjamin commented on src/Processors/Transforms/FilterTransform.cpp:
```diff
@@ -137,7 +153,11 @@ void FilterTransform::doTransform(Chunk & chunk)
     constant_filter_description = ConstantFilterDescription(*filter_column);
 
     if (constant_filter_description.always_false)
+    {
```
I think the next condition (`if (constant_filter_description.always_true)`) is redundant. This is not the result of your PR, but with your changes it became confusing, because you are rightfully not handling the profile events there. Could you please remove that?

onkar replied:
Removed the `if (constant_filter_description.always_true)` block altogether.

antaljanosbenjamin commented on tests/queries/0_stateless/03362_filter_transform_profile_events.sql:
```diff
@@ -0,0 +1,2 @@
+SELECT * FROM system.numbers WHERE number % 2 = 0 LIMIT 100;
+SELECT count() = 2 AS assert_exists FROM system.events WHERE name IN ('FilterTransformPassedRows', 'FilterTransformPassedBytes') HAVING assert_exists;
```
I think a better check would be to test the value of the mentioned profile events. You can use the `ProfileEvents` of the `system.query_log` system table ([docs](https://clickhouse.com/docs/operations/system-tables/query_log)). If you search for `ProfileEvents` or `system.query_log` in the tests then you will find nice examples how to use it.

onkar replied:
Added another test to look at the values of profile events.

onkar replied:
@antaljanosbenjamin I added the following test but it does not generate deterministic results. As seen in my commit. So, I will remove it. I looked into other ways but there is no way to make sure that same values are returned for `FilterTransformPassedBytes` and `FilterTransformPassedRows` every time, thus making this test flaky.

```
SYSTEM FLUSH LOGS;
SET log_queries = 1;
SELECT ProfileEvents['FilterTransformPassedRows'] AS passed_rows, ProfileEvents['FilterTransformPassedBytes'] AS passed_bytes FROM system.query_log WHERE type = 'QueryFinish' AND current_database = currentDatabase() ORDER BY event_time DESC LIMIT 1;
```

antaljanosbenjamin replied:
I would assume at least `FilterTransformPassedRows` should be deterministic. What causes that not to be? For bytes, I also don't see what cause the flakyness. Maybe I am not seeing everything though, so there might be a good explanation.

onkar replied:
The flakiness is caused because `system.query_log`'s flushing happens asynchronously. Using `SYSTEM FLUSH LOGS` does not guarantee that latest query stats will be immediately available. This is why I would like to keep the existing test as is. Additionally, other queries might have finished processing and `FilterTransformPassedRows`, `FilterTransformPassedBytes` values reflect their values in the result making it inconsistent.

onkar replied:
@antaljanosbenjamin thank you for this suggestion. This is working now. I have updated the diff with the extra test. Can we please merge this PR?

onkar replied:
Hello @antaljanosbenjamin, can this PR be merged?

onkar committed:
Address remaining review feedback

onkar committed:
Fix test failures

onkar committed:
Stabilize stateless test

onkar committed:
Add test with deterministic output

onkar committed:
Add reference file

antaljanosbenjamin commented:
Sorry for the late answer, I was busy. The PR looks good. Could you please resolve the conflicts?

The test failures seems to be unrelated.

onkar committed:
Merge branch 'master' into onkar/profile-event

onkar commented:
> Sorry for the late answer, I was busy. The PR looks good. Could you please resolve the conflicts?
> 
> The test failures seems to be unrelated.

Thanks for the re-review @antaljanosbenjamin. I have resolved the conflicts. Please take a look and merge if everything looks good.

antaljanosbenjamin commented:
`CH Inc sync` failed for unrelated reason, merging this.

clickgapai commented:
Hi @onkar @antaljanosbenjamin — while reviewing this PR I found the following:

- **Bug (P2):** FilterTransformPassedBytes counts pre-filter column sizes at line 247 — https://github.com/ClickHouse/ClickHouse/issues/101737

Happy to discuss — close anything that's wrong or already addressed.
